python/ray/tune/schedulers/resource_changing_scheduler.py

```diff
@@ -0,0 +1,222 @@
+import logging
+from typing import Dict, Optional, Union, Callable
+
+import pickle
+import warnings
+
+from ray.tune import trial_runner
+from ray.tune.resources import Resources
+from ray.tune.schedulers.trial_scheduler import FIFOScheduler, TrialScheduler
+from ray.tune.trial import Trial
+from ray.tune.utils.placement_groups import PlacementGroupFactory
+
+logger = logging.getLogger(__name__)
+
+
+class ResourceChangingScheduler(TrialScheduler):
+    """A utility scheduler to dynamically change resources of live trials.
+
+    Experimental. API may change in future releases.
+
+    The ResourceChangingScheduler works by wrapping around any other
+    scheduler and adjusting the resource requirements of live trials
+    in response to the decisions of the wrapped scheduler
+    through a user-specified `resources_allocation_function`. An example
+    of such a function can be found in
+    `tune/examples/xgboost_dynamic_resources_example.py`.
+
+    Only supports the Trainable (class) API. If resources of a trial are
```

Comment:
I'm wondering if we can find a way to support this in function trainables as well.

E.g.

```
def train(config):
    pass
    
train.resource_allocation_fn = distribute_cpu_gpu_evenly
```

or something like that. Or maybe

```
@tune.trainable(resource_allocation_fn=distribute_cpu_gpu_evenly)
def train(config):
    pass
```

And then pass this property in the function API wrapper.

Probably out of scope for this PR but maybe an idea for a follow up PR?

Reply:
I'll make an issue for that once this is merged